website/project/views/comment.py

```diff
@@ -13,106 +14,169 @@
 
 from website import settings
 from website.notifications.emails import notify
+from website.notifications.constants import PROVIDERS
 from website.filters import gravatar
 from website.models import Guid, Comment
+from website.addons.base import GuidFile
 from website.project.decorators import must_be_contributor_or_public
 from datetime import datetime
 from website.project.model import has_anonymous_link
+from website.project.views.node import _view_project, n_unread_comments
+from website.addons.figshare.exceptions import FigshareIsDraftError
 
 
-def resolve_target(node, guid):
+@must_be_contributor_or_public
+def view_comments_project(auth, **kwargs):
+    """
+    Returns information needed to get comments for the total discussion page
+    """
+
+    node = kwargs['node'] or kwargs['project']
+    page = request.args.get('page', None)
+    root_id = node._id if page == Comment.OVERVIEW else None
+
+    if page:
+        _update_comments_timestamp(auth, node, page=page, root_id=root_id)
+
+    ret = {
+        'comment_target': page or 'total',
+        'comment_target_id': root_id
+    }
+    ret.update(_view_project(node, auth, primary=True, check_files=(not page)))
+    return ret
+
+
+@must_be_contributor_or_public
+def view_comments_single(auth, **kwargs):
+    """
+    Returns information needed to get a single comment and its replies
+    """
+    node = kwargs['node'] or kwargs['project']
+    comment = kwargs_to_comment(kwargs)
+    serialized_comment = serialize_comment(comment, auth)
 
+    from website.addons.wiki.model import NodeWikiPage
+
+    ret = {
+        'comment': serialized_comment,
+        'comment_target': serialized_comment['page'],
+        'comment_target_id': comment.root_target.page_name
+        if isinstance(comment.root_target, NodeWikiPage)
+        else comment.root_target._id
+    }
+    ret.update(_view_project(node, auth, primary=True))
+    return ret
+
+
+def resolve_target(node, page, guid):
     if not guid:
         return node
     target = Guid.load(guid)
     if target is None:
+        if page == Comment.WIKI:
+            return node.get_wiki_page(guid, 1)
         raise HTTPError(http.BAD_REQUEST)
     return target.referent
 
 
-def collect_discussion(target, users=None):
-
-    users = users or collections.defaultdict(list)
-    for comment in getattr(target, 'commented', []):
-        if not comment.is_deleted:
-            users[comment.user].append(comment)
-        collect_discussion(comment, users=users)
+def collect_discussion(target, users):
```

Comment:
The `users` variable is misleading, as it's not a list of users--it's a dictionary. Can you improve the variable name (also changing the `users` variable on ~L93).